packages/service-provider-core/src/whatwg-url.ts

```diff
@@ -6,11 +6,21 @@ if (typeof require('util').TextDecoder !== 'function' || typeof require('util').
   Object.assign(require('util'), textEncodingPolyfill());
 }
 
-import { URL as WhatWGURL } from 'whatwg-url';
-import { URL as NodeURL } from 'url';
-export const URL = NodeURL ?? (WhatWGURL as typeof NodeURL);
+let URL: typeof import('url').URL;
 
-export function textEncodingPolyfill(): any {
+// Easiest way to get global `this` in any environment
+// eslint-disable-next-line no-new-func
+const globalThis = new Function('return this')();
+
+// Dynamically requiring URL should allow us to skip imports that potentially
+// can break browser runtimes
+if ('URL' in globalThis) {
+  URL = globalThis.URL;
+} else {
+  URL = require('url').URL ?? require('whatwg-url').URL;
+}
+
+function textEncodingPolyfill(): any {
```

Comment:
I think also this thing here is not super clear what is for, probably a comment would help

Reply:
The `whatwg-url` polyfill needs it (since neither `URL` nor `TextEncoder`/`TextDecoder` are available in the java-shell)

Reply:
Ah, sorry, it's just my changes making the diff confusing I think. There is a nice comment on top by @addaleax explaining the `textEncodingPolyfill`:

https://github.com/mongodb-js/mongosh/blob/641a193fb6ceda5c0cb94b76d971f0e9f5854bab/packages/service-provider-core/src/whatwg-url.ts#L1-L4

Reply:
Is a matter of taste but I find the ironic comment there (`, because it's so much fun,`) a bit out of place, although I appreciate it 😆

Reply:
> Is a matter of taste but I find the ironic comment there , because it's so much fun, a bit out of place, although I appreciate it 😆.

I take it back, is all right, i thought it was referring to the java shell implementation, but actually is just the java runtime that is like that  .. sorry again, i should learn to read before doing PRs ..

Reply:
oh, yeah, sorry if this came across as directed at anyone! feel free to remove it if it’s ambiguous. It was admittedly a bit frustrating to get everything to work, and that may have seeped through here, but since `URL` is not language feature, but a browser/Node.js one, there’s really nothing wrong here, just something that we have to work around